torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1150,11 +1158,101 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
   // threads dying due to aborted communicator and raising a SIGABRT
   std::string abortReason = c10::str("Process Group destroyed on rank ", rank_);
   abort(abortReason);
+
+  // We need to wait for abort to finish before we can safely shut down
+  // heartbeat monitoring thread.
+  terminateHeartbeatMonitorThread_.store(true);
+  monitorWakeUpCV_.notify_one();
+#ifdef ENABLE_NCCL_ERROR_CHECKING
+  if (ncclHeartbeatMonitorThread_.joinable()) {
+    ncclHeartbeatMonitorThread_.join();
+  }
+#endif
+}
+
+void dump_debugging_info() {
+  LOG(ERROR)
+      << "No PGNCCL's watchdog heartbeat detected, so we are dumping debug info.";
+  if (parseEnvVarIntDefault("TORCH_NCCL_TRACE_BUFFER_SIZE", 0) > 0) {
+    // TODO: Find the right and proper way to dump the debug info.
+    // We cannot print out debugging info directly.
+    LOG(ERROR) << "nccl_trace: " << dump_nccl_trace();
+  }
+}
+
+void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
+  // Logging with `FATAL`, after errMsg printed, it calls `std::abort()`
+  // to terminate the program execution.
+  LOG(FATAL) << errMsg;
+}
+
+void ProcessGroupNCCL::heartbeatMonitor() {
+  uint64_t heartBeatCounter = 0ULL;
+  while (true) {
+    // This won't have any lock since this lock is only used here.
+    // Please be aware that mutex `monitorMutex_` should not be used
+    // somewhere else to avoid the deadlock.
+    std::unique_lock<std::mutex> lock(monitorMutex_);
+    if (monitorWakeUpCV_.wait_for(
+            lock, std::chrono::seconds(heartbeatTimeoutInSec_), [&] {
+              return terminateHeartbeatMonitorThread_.load();
+            })) {
+      // For the normal complete or user interception, monitorWakeUpCV_
+      // will get notified, we early return and exit heartbeatMonitor.
+      return;
+    } else {
+      auto heartbeat = heartbeat_;
+      if (heartbeat > heartBeatCounter) {
```

Comment:
i also picked on this before. junjie made the int bigger and claimed we'll never hit it.  I prefer we handle the wraparound too.  (but if we do the wraparound wrong, its worse than if we don't do it at all, and i think it could be  a bit complicated to do the wraparound).

Reply:
`heartbeat_` is unsigned long, so quote my reply to one of Will's previous comment:

> so the range is 0 to 4,294,967,295. If we assume the heartbeat is sent every second, then it is 49K days, roughly 136 years.. That should be sufficient for any types of training jobs.

Reply:
OK, I think I will just change `>` to `!=` since heartBeat can be assumed monotonically increasing.

Reply:
And we will keep unsigned long for now.